File: java/MergeSort.java
import java.util.ArrayList;
import java.util.Iterator;

public class MergeSort
{

		public ArrayList<Integer> mergeSort(final ArrayList<Integer> vals)
		{
			return divide(vals, 0, vals.size()-1);
		}

		public ArrayList<Integer> divide(final ArrayList<Integer> vals, final int start, final int end){
			if(start > end)
			{
				return null;
			}
			else if(start == end)
			{
				ArrayList<Integer> v = new ArrayList<Integer>(1);
				v.add(vals.get(start));
				return v;
			}
			int mid = (start + end) / 2;
			ArrayList<Integer> first = divide(vals, start, mid);
			ArrayList<Integer> second = divide(vals, mid + 1, end);
			return merge(first, second);
		}

		public ArrayList<Integer> merge(final ArrayList<Integer> list1, final ArrayList<Integer> list2)
		{
			ArrayList<Integer> ret = new ArrayList<Integer>(list1.size() + list2.size());
			Iterator<int> list1_it = list1.iterator();
			Iterator<int> list2_it = list2.iterator();
			int val1 = list1_it.next();
			int val2 = list2_it.next();
			while(list1_it.hasNext() && list2_it.hasNext())
			{
				if(val1 <= val2)
				{
					ret.add(val1);
					val1 = list1_it.next();
				}
				else
				{
					ret.add(val2);
					val2 = list2_it.next();
				}
			}

			while(list1_it.hasNext())
			{
				ret.add(val1);
				val1 = list1_it.next();
			}

			while(list2_it.hasNext())
			{
				ret.add(val2);
				val2 = list2_it.next();
			}
			return ret;
		}

}

File: python/merge_sort.py
class Solution(object):
	def merge_sort(self, vals):
		return self.divide(vals, 0, len(vals)-1)

	def divide(self, vals, start, end):
		if start > end:
			return None
		if start == end:
			return [vals[start]]
		mid = (start + end)/2
		first = self.divide(vals, start, mid)
		second = self.divide(vals, mid+1, end)
		return self.merge(first, second)

	def merge(self, list1, list2):
		ret = [None for i in xrange(len(list1) + len(list2))]
		i = j = r = 0
		while (i < len(list1)) and (j < len(list2)):
			if list1[i] <= list2[j]:
				ret[r] = list1[i]
				i += 1
			else:
				ret[r] = list2[j]
				j += 1
			r += 1

		while i < len(list1):
			ret[r] = list1[i]
			i += 1
			r += 1

		while j < len(list2):
			ret[r] = list2[j]
			j += 1
			r += 1
		return ret

if __name__ == "__main__":
	# x = [1,3]
	# y = [2]
	z = Solution()
	# print z.merge(x,y)

	import random
	x = [random.randint(1,500) for i in xrange(random.randint(1,50))]
	print x
	print z.merge_sort(x)